include/oneapi/dpl/internal/distributed_ranges_impl/sp/algorithms/for_each.hpp

```diff
@@ -0,0 +1,94 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <sycl/sycl.hpp>
+
+#include "../../detail/sycl_utils.hpp"
+#include "../detail.hpp"
+#include "../init.hpp"
+#include "../util.hpp"
+#include "../zip_view.hpp"
+#include "execution_policy.hpp"
+
+namespace oneapi::dpl::experimental::dr::sp
+{
+
+template <typename ExecutionPolicy, distributed_range R, typename Fn>
+void
+for_each(ExecutionPolicy&& policy, R&& r, Fn fn)
+{
+    static_assert( // currently only one policy supported
+        std::is_same_v<std::remove_cvref_t<ExecutionPolicy>, sycl_device_collection>);
+
+    std::vector<sycl::event> events;
+
+    for (auto&& segment : ranges::segments(r))
+    {
+        auto&& q = __detail::queue(ranges::rank(segment));
+
+        assert(stdrng::distance(segment) > 0);
+
+        auto local_segment = __detail::local(segment);
+
+        auto first = stdrng::begin(local_segment);
+
+        auto event = dr::__detail::parallel_for(q, sycl::range<>(stdrng::distance(local_segment)),
+                                                [=](auto idx) { fn(*(first + idx)); });
+        events.emplace_back(event);
+    }
+    __detail::wait(events);
+}
+
+template <typename ExecutionPolicy, distributed_iterator Iter, typename Fn>
+void
+for_each(ExecutionPolicy&& policy, Iter begin, Iter end, Fn fn)
+{
+    for_each(std::forward<ExecutionPolicy>(policy), stdrng::subrange(begin, end), fn);
+}
+
+template <distributed_range R, typename Fn>
+void
+for_each(R&& r, Fn fn)
```

Comment:
1) That re-call of `for_each` w/o policy to `for_each(par_unseq)` is done by design? Is the some explanation and statement in the documentation?
2) According to the static_assert (line 34) above `par_unseq` is not currently supported...
3) Taking into account 1) and 2) does whether it make sense to release non-policy `for_each` versions at all?

Reply:
These algorithms are restricted by distributed range/iterator concepts. It's a question if we want to keep it this way in the long run, but let's keep the approach as-is for now.

Reply:
> inline sycl_device_collection par_unseq;

It totally confuses any C++ developers which are aware of C+17 execution policies.
(and it is a potential name conflict...)
I would suggest renaming one in something proper name.

Reply:
Can you explain what's problematic about using the name `par_unseq` as a catch-all execution policy to execute across multiple devices?

From my understanding of the spec, [`par_unseq` requires](https://eel.is/c++draft/algorithms.parallel#exec-7):

- Execution in "unordered fashion" in "unspecified threads of execution," and
- These threads must provide "weakly parallel forward progress guarantees."

From what I understand, threads in SYCL have weakly parallel forward progress guarantees.  Is it the "threads of execution" part that would be violated?

Other libraries [like Nvidia's stdpar](https://developer.nvidia.com/blog/accelerating-standard-c-with-gpus-using-stdpar/) use `par_unseq` to mean "execute in parallel, including on GPUs if that's where the data is."  That's what we're doing here, and it seems like a reasonable choice to me.  Is there a convention in oneDPL that `par_unseq` is CPU only, or is there another good reason to not use `par_unseq` as the "sane default" that includes potentially running on a GPU if that's where the data is resident?